src/routers/alpha-router/functions/get-candidate-pools.ts

```diff
@@ -357,149 +278,265 @@ export async function getV3CandidatePools({
+    if (
+      topByBaseWithTokenIn.length >= topNWithBaseToken &&
+      topByBaseWithTokenOut.length >= topNWithBaseToken &&
+      topByEthQuoteTokenPool.length >= 2 &&
+      topByTVL.length >= topN &&
+      topByTVLUsingTokenIn.length >= topNTokenInOut &&
+      topByTVLUsingTokenOut.length >= topNTokenInOut
+    ) {
+      // We have satisfied all the heuristics, so we can stop.
+      break;
+    }
 
-  addToAddressSet(topByTVL);
+    if (
+      topByBaseWithTokenIn.length < topNWithBaseToken &&
+      (
+        (baseTokensAddresses.includes(subgraphPool.token0.id) && subgraphPool.token1.id == tokenInAddress) ||
+        (baseTokensAddresses.includes(subgraphPool.token1.id) && subgraphPool.token0.id == tokenInAddress)
+      )
+    ) {
+      topByBaseWithTokenIn.push(subgraphPool);
+      // This pool has been added to the heuristic, continue with the next pool
+      continue;
+    }
 
-  const topByTVLUsingTokenIn = _(subgraphPoolsSorted)
-    .filter((subgraphPool) => {
-      return (
-        !poolAddressesSoFar.has(subgraphPool.id) &&
-        (subgraphPool.token0.id == tokenInAddress ||
-          subgraphPool.token1.id == tokenInAddress)
-      );
-    })
-    .slice(0, topNTokenInOut)
-    .value();
+    if (
+      topByBaseWithTokenOut.length < topNWithBaseToken &&
+      (
+        (baseTokensAddresses.includes(subgraphPool.token0.id) && subgraphPool.token1.id == tokenOutAddress) ||
+        (baseTokensAddresses.includes(subgraphPool.token1.id) && subgraphPool.token0.id == tokenOutAddress)
+      )
+    ) {
+      topByBaseWithTokenOut.push(subgraphPool);
+      continue;
+    }
 
-  addToAddressSet(topByTVLUsingTokenIn);
+    // Main reason we need this is for gas estimates, only needed if token out is not native.
+    // We don't check the seen address set because if we've already added pools for getting native quotes
+    // there's no need to add more.
+    if (
+      topByEthQuoteTokenPool.length < 2 &&
+      (
+        (
+          WRAPPED_NATIVE_CURRENCY[chainId]?.symbol == WRAPPED_NATIVE_CURRENCY[ChainId.MAINNET]?.symbol &&
+          tokenOut.symbol != 'WETH' &&
+          tokenOut.symbol != 'WETH9' &&
+          tokenOut.symbol != 'ETH'
+        ) ||
+        (
+          WRAPPED_NATIVE_CURRENCY[chainId]?.symbol == WMATIC_POLYGON.symbol &&
+          tokenOut.symbol != 'MATIC' &&
+          tokenOut.symbol != 'WMATIC'
+        )
+      ) &&
+      (
+        routeType === TradeType.EXACT_INPUT && (
+          (subgraphPool.token0.id == wrappedNativeAddress && subgraphPool.token1.id == tokenOutAddress) ||
+          (subgraphPool.token1.id == wrappedNativeAddress && subgraphPool.token0.id == tokenOutAddress)
+        ) ||
+        routeType === TradeType.EXACT_OUTPUT && (
+          (subgraphPool.token0.id == wrappedNativeAddress && subgraphPool.token1.id == tokenInAddress) ||
+          (subgraphPool.token1.id == wrappedNativeAddress && subgraphPool.token0.id == tokenInAddress)
+        )
```

Comment:
nit: ik this was from existing code but this conditional is really hard to follow, would support breaking it up into helper functions like `poolContainsNative` etc.

Reply:
yeah it is definitely hard to follow, I think a lot of these conditions will ahve to change at some point, but my main focus for this branch was to remove the multiple loops.

I can create a helper function, but it is also confusing because all the v2, v3 and mixed candidate pool functions are inside this `get-candidate-pools.ts` file.

I think there's way more room for improvement in here, and I would defer a bit on further changes on this branch until we spend the time refactoring

Reply:
yeah makes sense